### ISSUE TITLE
Provide "native" Extension support

### ISSUE THREAD

**Issue by phiggins42** [yabwe/medium-editor#519]
tl;dr - There should be a core `Extension` contructor, providing the baseline API needed to implement an editor extension. Buttons should extend this Extension, defining button-interaction things. Internal extensions should use a normalized API for interacting with the editor. The extension mechanism should be exposed to end-users wishing to implement their own extensions. 

---

* the `statics` portion of the place where these defaults are exposed is kind of overkill … people are used to the overloaded ctor/namespace, provided the usage is documented well enough, and providing extendable base class things for various use cases (pure extension/behavior/event-listener hook vs actual-toolbar-button) would make external extension that much easier and consistent.

---

Currently, as of HEAD -- one _can_ make an extension tweak like:

```
   var MyButton = function(options, instance){
       this.parent = true;
       this.name = "myButton";
       this.init(options, instance);
   }

   MyButton.prototype = {
       // stuff
   }

   var MyExtension = MediumEditor.util.derives(MediumEditor.statics.DefaultButton, MyButton);

   new MediumEditor(.., {
       extensions: {
           "myButton": new MyExtension()
       }
   });
```

feels kind of clunky. We're mostly concerned with linerar single-inheritance chains here, so a syntax like this is not unreasonable:

```
   var MyButton = MediumEditor.extensions.DefaultButton.extend({
       // stuff
   });

   new MediumEditor(..., {
       extensions: {
           "myButton": new MyButton()
       }
   });
```

Functionally the two forms are identical. The ticket from tonight #376, the code would be:

```
   FontSizeExtension = MediumEditor.Anchor.extend({
       init: function(){
           this.options = {
               // we hard-code/build up our ButtonsData type def
           }
       }
       // the 4-or-5 override functions, the rest is the same
       // you call the super with `MediumEditor.Anchor.prototype.thisFunction.apply(this, arguments);`
   });
```

The proposal would be something like:

   * provide MediumEditor.Extension base to extend from.
   * rename/deprecate MediumEditor.statics.* -> MediumEditor.*
   * expose PasteHandler in a similar fashion (allowing people to append/overwrite handling there)?
   * add MediumEditor.extend so that people can bulk extend the core API's without touching it.

```
   var Extension = function(options, instance){
       this.options = options || this.options || {};
       this.init(this.options, instance);
   }

   Extension.prototype = {
       // base Button API definition.
       // parent: true,
       // hasForm: false,
       // yadda yadda
   };

   Extension.extend = function(props){
       // magic that returns a new ctor inheriting Extension's 
       // prototype, mixing props over top and relaying this
       // .extend function, so the "subclass" can be similarly
       // extended.
   }
```

Then, anything that needs to be an Extension can just be:

```
   var MyButton = MediuEditor.Extension.extend({
       options:{
           // button style options, action/etc
       }
   });

   new MediumEditor(..., {
       extensions:{
           "myButton": new MyButton()
       }
   });
```

This doesn't _require_ the user to write extensions this way, just provides them a clean/simple way to write extensions or potentially override built-in extensions outright using built-in tooling provided by the editor for such a purpose (1st class citizens, defined/documented lifecycle API).

It keeps everything as simple as it was and opens the door to all kinds of hacking and extension.

**Comment by nmielnik** [comment]
* I'm definitely on board with a lot of this, I think adding an `.extend` method to the objects we want things to be derived from is a great idea.

Personally, I would be in favor of exposing these via `MediumEditor.extensions` like you did in an earlier example, as opposed to on the MediumEditor object itself.

```js
MediumEditor.extensions.Anchor
MediumEditor.extensions.AnchorPreview
MediumEditor.extensions.FontSize
MediumEditor.extensions.DefaultButton
```

* If we wanted to get rid of statics altogether, we could move `ButtonsData` as a static member of `DefaultButton` or even `MediumEditor` itself.  I'm not sure how/if we should expose the Toolbar object.  It's being use in the test cases, but I would also be onboard with not exposing it at all for now.

* In regards to the PastHandler, we would need to do a bit more work to make that overrideable, it may be that we don't need to override, but just need to expose a few options for pasting to make it configurable.

* I'm less ok with exposing `MediumEditor.extend()`. It would be good to have at least 1 real use case for it to see if we can't account for most cases via options.  It's definitely easier for us to make tests around `options` and ensure they don't break, as opposed to having people use `extend()` and us break them without us really knowing it.  My $0.02 anyway.

**Comment by nmielnik** [comment]
@nchase and @daviferreira , would love your input on this.

Other folks who may have some valuable feedback: @secrettriangle @IonicaBizau @martinth @arcs- @orthes

**Comment by phiggins42** [comment]
> Personally, I would be in favor of exposing these via MediumEditor.extensions like you did in an earlier > example, as opposed to on the MediumEditor object itself.

I agree making a rename/move to use `.extensions = {}` would be suitable. Exposing the `ButtonDefaults?`on the MediumEditor makes a lot of sense, too. 

> I'm less ok with exposing MediumEditor.extend(). It would be good to have at least 1 real use case for it > to see if we can't account for most cases via options. It's definitely easier for us to make tests around 
> options and ensure they don't break, as opposed to having people use extend() and us break them 
> without us really knowing it. My $0.02 anyway.

fair enough, though it would be 1-loc for anyone to do it:

```
MediumEditor.extend = MediumEditor.util.extender;
```

After looking more at pasteHandler and Toolbar it seems there is a slight inconsistency in how ctor args are passed and used. they both are candidates for being "just extensions", but would require their existing ctor args to be re-worked ... I feel like the Toolbar behavior is so fundamental to custom UX and tweaking it _should_ be exposed, and treated as an "extension" like all the others (allowed to register the events it needs, draw a dom, all it needs to do it implement the base extension apis and more or less does already?)

**Comment by nmielnik** [comment]
Yeah, I've gone back and forth on Toolbar.  Since so much of the critical code for medium-editor lives in there, it doesn't seem like it should be an extension as much as should be an overrideable object that could be passed in during MediumEditor creation.  Not sure how you would feel about that kind of difference.

The story with PasteHandler was it began as a group of static methods (like Util), but at some point I decided to just make it an instantiated object so it could have better access to options, and so it would have a chance to attach event handlers.  I think the next logical step would be to make it an extension like you mentioned, and perhaps Placeholders should be as well.

**Comment by daviferreira** [comment]
I think we need to step out a bit and rethink MediumEditor architecture, maybe plan for a 5.0 release. I will elaborate when I ave that time, but nice discussion so far. I think this is the right moment for it, sice now we have finished with our first goal that was to eliminate the one big file problem we had.

I also don't like the way we pass options, or the editor instance, right now. That definitely needs to be revisited.

Would be really nice to hear from people who developed extensions, as Nate said.

**Comment by secrettriangle** [comment]
I am definitely on board with an extension architecture. The specifics don't matter too much as long as an extension author would have the tools necessary to get his job done. Would be great if *all* the buttons (bold/italic/etc) were implemented as an extension.

**Comment by IonicaBizau** [comment]
Nice work! :+1: Having such a feature would be awesome! :sparkles:

**Comment by phiggins42** [comment]
I made a gist, as my random thoughts expanded well beyond what would be acceptable as a comment. 

https://gist.github.com/phiggins42/a9dd79afb4fc6feb2ec1

**Comment by nmielnik** [comment]
This is done, everything (including the toolbar) is now an extension that you can `extend()`.  This is all now fully available in v5.0.0, and more details on extensions can be found in our [Wiki](https://github.com/yabwe/medium-editor/wiki)!
